evcache-core/src/main/java/net/spy/memcached/EVCacheMemcachedClient.java

```diff
@@ -124,62 +127,85 @@ private boolean isWrongKeyReturned(String original_key, String returned_key) {
-                } catch (Exception e) {
-                    log.error(e.getMessage(), e);
-                    rv.set(null, status);
-                }
-            }
+        final EVCacheOperationFuture<T> rv = new EVCacheOperationFuture<>(key, latch, new AtomicReference<T>(null), readTimeout.get(), executorService, client);
+        final DistributionSummary dataSizeDS = getDataSizeDistributionSummary(
+                EVCacheMetricsFactory.GET_OPERATION,
+                EVCacheMetricsFactory.READ,
+                EVCacheMetricsFactory.IPC_SIZE_INBOUND);
+
+        final Operation op = opFact.get(key, new GetOperation.Callback() {
+            private volatile boolean decodingData = false;
+            private volatile OperationStatus status = null;
+            private final AtomicReference<T> valueHolder = new AtomicReference<>(null);
 
             @SuppressWarnings("unchecked")
             public void gotData(String k, int flags, byte[] data) {
+                decodingData = true;
+                if (isWrongKeyReturned(key, k)) {
+                    latch.countDown();
+                    return;
+                }
 
-                if (isWrongKeyReturned(key, k)) return;
+                boolean shouldLog = log.isDebugEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName);
+                boolean alwaysSync = false;
 
-                if (log.isDebugEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName)) log.debug("Read data : key " + key + "; flags : " + flags + "; data : " + data);
+                if (shouldLog) log.debug("Read data : key {}; flags : {}; data : {}", key, flags, data);
                 if (data != null)  {
-                    if (log.isDebugEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName)) log.debug("Key : " + key + "; val size : " + data.length);
-                    getDataSizeDistributionSummary(EVCacheMetricsFactory.GET_OPERATION, EVCacheMetricsFactory.READ, EVCacheMetricsFactory.IPC_SIZE_INBOUND).record(data.length);
-                    if (tc == null) {
-                        if (tcService == null) {
-                            log.error("tcService is null, will not be able to decode");
-                            throw new RuntimeException("TranscoderSevice is null. Not able to decode");
-                        } else {
-                            final Transcoder<T> t = (Transcoder<T>) getTranscoder();
-                            val = tcService.decode(t, new CachedData(flags, data, t.getMaxSize()));
-                        }
+
+                    if (shouldLog) log.debug("Key : {}; val size : {}", key, data.length);
+                    dataSizeDS.record(data.length);
+                    Transcoder<T> transcoder = (tc == null) ? (Transcoder<T>) getTranscoder() : tc;
+                    if (tcService == null) {
+                        log.error("tcService is null, will not be able to decode");
```

Comment:
should we do latch.countDown() here?

Reply:
fixed in 2b60f17, I restructured it too. Hopefully reduce chance of bugs like this.